Getting issues using filters

nmadruga opened the pull request:
To test this branch, you can either:

1) (Using the terminal) Run it locally:
- Update the file `.env.example` giving the REPO, OWNER and WEBHOOK_SECRET (which can be your own token from Github user - with access to the repo). Rename the file to `.env`
-  Give parameters as expected. i.e: `node changesLog.js -m 25 -f "Bug: Production, Feature: new, Feature: existing"` where -m 25 is the Milestone number for mSupply version 4.09.00.

2) (Using the terminal) Run it from omsupply-bot: - Require some changes on the branch `36-generate-change-log-for-new-releases` first...

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
-function commaSeparatedList(value, dummyPrevious) {
+function commaSeparatedList(value) {
   return value.split(',');
 }
```
Any reason for ES5 function synxtax? IMO would be nicer to keep consistent with other repos and use arrow syntax?

```suggestion
const commaSeparatedList = value => value.split(',');
```
```

nmadruga replied:
I agree. I was only avoiding installing packages to convert things from ES6, but that is accepted for ES5.

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
 const commander = require('commander');
-const { getIssuesByMilestone } = require('./changesLogGenerator');
+const fs = require('fs');
+const { fetchIssuesInMilestone, generateChangesLog } = require('./helpers');
```
IMO we should stick to ES6 `import` notation. Maybe better for separate "babel" issue?

nmadruga replied:
Yes. I'm not dealing with packages for now.

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+  .option('-i, --includeAll <flag>', 'will include a section with issues done for all clients (Default: false)')
+  .option('-s, --state <state>', 'should use issues with state: closed, open, all (Default: closed)', checkValidStates)
+  .option('-f, --filters <labels>', 'comma separated labels to filter and group by issues', commaSeparatedList)
+  .option('-d, --duplicate <allow>', 'should allow duplicated issues in different groups (Default: true)')
```
Nit: I think would be good to define these option strings as constants?

nmadruga replied:
I don't agree. I feel that this way, since that's the only place this is used it looks more easy to read like this. Please, let me know if you are strong about this and we can discuss the idea.

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+function checkValidStates(value) {
+  return (value === 'closed') || (value === 'open') || (value === 'all') ? value : undefined;
+}
+
```
I think this function is a bit confusing? `checkX` should probably return a `boolean` value? Maybe better to return `true` or `false` and let the caller handle how to proceed?

nmadruga replied:
Agreed. I'll make this change now. Thanks!

nmadruga replied:
Oh, actually the option function expects a callback that will return the value. I'll change callback name to validateStatus instead.

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
   return value.split(',');
 }
 
+function checkValidStates(value) {
```
Not sure what this function does? Maybe rename to, e.g. `isValidLabel`?

nmadruga replied:
It needs to return the value (if valid) so I changed only the method name to `validateState`

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
 }
 
+function checkValidStates(value) {
+  return (value === 'closed') || (value === 'open') || (value === 'all') ? value : undefined;
```
IMO should avoid explicitly returning `undefined` from a function (suggest using `null` instead). My thinking is that it's probably good practice to keep `undefined` reserved only for when a function doesn't return anything.

nmadruga replied:
No problem

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
   .parse(process.argv);
 
+let groupedIssues = [];
+const filters = commander.filters;
```
```suggestion
const { filters } = commander;
```

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+  state: commander.state ? commander.state : 'closed',
+  customer: commander.customer ? commander.customer : undefined,
+  duplicateIssues: commander.duplicate ? commander.duplicate : true,
+  includeIssueForAll: commander.includeAll ? commander.includeAll : true
```
I think a bit nicer to use destructuring with defaults here:

```suggestion
  const { milestone, state = 'closed', customer = null, duplicate = true, includeAll: includeIssueForAll = true } = commander;
  const params = { milestone, state, customer, duplicate, includeIssueForAll }
```

Thoughts?

nmadruga replied:
Wow! That looks great. Thanks Will! I just didn't use because I had no idea it could be done that way. I'll add and test if will work the same.

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
-  filters: commander.filters,
-  groups: commander.groups
+  state: commander.state ? commander.state : 'closed',
+  customer: commander.customer ? commander.customer : undefined,
```
Is it needed to not have the key defined here? IMO best practice is to define all keys consistently (even if `null`), and pass off `null` checking to functions which consume the object.

nmadruga replied:
I see... I'll have a look. I haven't put to much thought in it before.

nmadruga replied:
Done

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+  state: commander.state ? commander.state : 'closed',
+  customer: commander.customer ? commander.customer : undefined,
+  duplicateIssues: commander.duplicate ? commander.duplicate : true,
+  includeIssueForAll: commander.includeAll ? commander.includeAll : true
```
Do we need both identifiers: `includeAll`, `includeIssueForAll` ?

nmadruga replied:
Ahh no. We don't I'll keep the longest one because I think is more explanatory

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
-console.log(`Duplicate issues (in different groups) set to: ${params.duplicate}.`)
-if (params.filters !== undefined) console.log(`Filters: ${params.filters}`);
-if (params.groups !== undefined) console.log(`Groups: ${params.groups}`);
+async function fetchIssues() {
```
```suggestion
const fetchIssues = async () => {
```

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+async function fetchIssues() {
+  console.log(`\n\nFetching issues in Milestone ${params.milestone}...\n`);
+  if(filters) {
+    groupedIssues = await fetchIssuesInMilestone(octokit, params, filters);
```
IMO no need to use method overloading here. Can make it a bit tidier by handing off parameter checking to the function implementation (i.e. the function body `null` checks `filters`, rather than the caller). Can still use conditional for switching on the log statements.

nmadruga replied:
Done. Totally agree

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    groupedIssues = await fetchIssuesInMilestone(octokit, params, filters);
+    
+    if (groupedIssues.length > 0) {
+      console.log(`\nFinished fetching issues by filters in Milestone ${params.milestone}!\n`);
```
Can these log strings be defined as constants?

nmadruga replied:
Yes. I'll make these change along with the few next ones and make a single commit.

nmadruga replied:
The changes are included, although I don't really like having to pass `params` in every call to logs. We need a better solution to this. Help!

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    
+    if (groupedIssues.length > 0) {
+      console.log(`\nFinished fecthing issues in Milestone ${params.milestone}!\n`);
+      console.log('Started creating Changes lof for issues in single block...\n');
```
Typo: `'log'`.

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    groupedIssues = await fetchIssuesInMilestone(octokit, params);
+    
+    if (groupedIssues.length > 0) {
+      console.log(`\nFinished fecthing issues in Milestone ${params.milestone}!\n`);
```
Typo: `'fetching'`.

wlthomson replied:
```suggestion
      console.log(`\nFetched issues assigned to milestone: ${params.milestone}!\n`);
```

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    
+    if (groupedIssues.length > 0) {
+      console.log(`\nFinished fetching issues by filters in Milestone ${params.milestone}!\n`);
+      console.log(`Duplicate issues (in different groups) set to: ${params.duplicateIssues}.\n`);
```
```suggestion
      console.log(`Duplicate issues are ${params.duplicateIssues ? 'enabled' : 'disabled'}.\n`);
```

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    
+    if (groupedIssues.length > 0) {
+      console.log(`\nFinished fetching issues by filters in Milestone ${params.milestone}!\n`);
+      console.log(`Duplicate issues (in different groups) set to: ${params.duplicateIssues}.\n`);
```
I think can remove the `'in different groups'` explanation? IMO good to keep the logs as concise as possible, and my thinking is that we should expect new users to read up on the `duplicateIssues` flag (using `-h` or docs) before running the changelog?

nmadruga replied:
Cool. I will change that

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    if (groupedIssues.length > 0) {
+      console.log(`\nFinished fetching issues by filters in Milestone ${params.milestone}!\n`);
+      console.log(`Duplicate issues (in different groups) set to: ${params.duplicateIssues}.\n`);
+      console.log(`Started creating Changes log for issues based on Groups: ${filters}...\n`);
```
I think this log should be moved to the function where the changelog is actually created?

nmadruga replied:
For this one I'm disagreeing. I think here is the best because: 
1. That's where we use for our use (and not probot) 
2. Keeps all logs where is the `client` method. And helpers are just busy with logic.

But I can be wrong, I just don't agree.

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    
+    if (groupedIssues.length > 0) {
+      console.log(`\nFinished fecthing issues in Milestone ${params.milestone}!\n`);
+      console.log('Started creating Changes lof for issues in single block...\n');
```
As above, think this log should be moved into the block of code where the change log is created.

nmadruga replied:
The code was changed to be included in same case

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    
+    if (groupedIssues.length > 0) {
+      console.log(`\nFinished fecthing issues in Milestone ${params.milestone}!\n`);
+      console.log('Started creating Changes lof for issues in single block...\n');
```
Question: what does `'in single block'` mean?

nmadruga replied:
Not using groups (i.e: ### Features, ### Bug fixes) those are groups.
I'll change this description to: "Creating Changes-log for issues without groups."

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+}
 
-printIssuesFromMilestone()
+async function asyncGenerateChangesLog() {
```
```suggestion
const asyncGenerateChangesLog = async () => {
```

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
 
-printIssuesFromMilestone()
+async function asyncGenerateChangesLog() {
+  if(!octokit.asyncTryToken()) {
```
I think it's a good pattern to avoid evaluating methods directly in conditionals? Mainly a style thing, but IMO also makes debugging easier. 

```suggestion
  const await isValidToken = octokit.asyncTryToken();
  if (!isValidToken) {
```

wlthomson replied:
Does this need an `await`?

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
-printIssuesFromMilestone()
+async function asyncGenerateChangesLog() {
+  if(!octokit.asyncTryToken()) {
+    console.log('The token is not valid!');
```
```suggestion
    console.log('Invalid GitHub authentication token!');
```

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    return;
+  }
+
+  if(await fetchIssues()) {
```
Same as above re. conditionals evaluating method calls.

nmadruga replied:
Will deal with it and next comment in the same commit

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+    return;
+  }
+
+  if(await fetchIssues()) {
```
IMO would be a good idea to return an object here with all the parameters needed (e.g. `groupedIssues`). Think that would be better than using globals, which are a bit dangerous and can make the code a bit confusing to read?

nmadruga replied:
Yes, that's reasonable. I'll make this change and test. Thank you

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+
+  if(await fetchIssues()) {
+    console.log('\nSuccessful fetch!');
+    const changesLog = generateChangesLog(groupedIssues, params);
```
Question: is there a reason for using `changesLog` (instead of `changeLog`)?

nmadruga replied:
Ah. For me... It seemed correct. But now I don't know. In English do we use changeLog even if there are several changes? If that's the case I might just rename all files and strings (in the end of this review). @wlthomson Just need your input in this one, please.

mark-prins replied:
English huh? 🤷‍♂ 

*changesLog* does make more logical sense - because we are talking about more than one change, so it should be plural. But.. whoever started calling it a `changeLog` went with singular on the change.

[wikipedia](https://en.wikipedia.org/wiki/Changelog) doesn't say who named it that way - just says that common usage is 'changeLog'

nmadruga replied:
Thank you @mark-prins I'll make this change at the end so the file renaming won't be a problem while I make all the changes for the review.

wlthomson commented on changesLog.js:
```diff
@@ -1,34 +1,75 @@
+  if(await fetchIssues()) {
+    console.log('\nSuccessful fetch!');
+    const changesLog = generateChangesLog(groupedIssues, params);
+    fs.writeFile('changesLog.txt', changesLog, (err) => {
```
If not a user param, I think output file should also be a constant?

nmadruga replied:
Done

wlthomson commented on constants/index.js:
```diff
@@ -0,0 +1,8 @@
+const { prefixes, getTitle, checkSomeLabelEquals, checkSomeLabelHasPrefix } = require('./labels');
```
As above, suggest using `import`.

wlthomson commented on constants/index.js:
```diff
@@ -0,0 +1,8 @@
+const { prefixes, getTitle, checkSomeLabelEquals, checkSomeLabelHasPrefix } = require('./labels');
+
+module.exports = {
```
If using `import`, can use `export` here.

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+
+const prefixes = {
```
I think for consistency good to make all constant identifiers caps (but not sure `PREFIX` is the best naming choice, see below):

```suggestion
const PREFIXES = {
```

nmadruga replied:
Done

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+
+const prefixes = {
```
I think `prefix` is correct, but since the prefixes are for grouping labels, maybe something like:

```suggestion
const LABEL_GROUPS = {
```

wlthomson replied:
I think even though in the `labels.js`, might be good to include `"LABEL"` in the label constant identifiers?

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+    FEATURE: 'Feature'
+};
+
+const bugTypes = {
```
```suggestion
const LABEL_TYPES_BUG = {
```

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+    DEVELOPMENT: 'development'
+}
+
+const featureTypes = {
```
```suggestion
const LABEL_TYPES_FEATURE = {
```

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+function getType (label) {
+    const labelType = (label || '').split(':');
+    return (labelType.length > 1) ? labelType[1].trimLeft() : '';
+}
```
Think these two functions can be combined:

```suggestion
const parseLabel = label => {
  if (!label) return { group: null, type: null };
  const [group = '', type= ''] = label.split(';');
  return {
    group: group.trimRight(),
    type: type.trimLeft(),
  };
}
```

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+    return (labelType.length > 1) ? labelType[1].trimLeft() : '';
+}
+
+function getTitle (label) {
```
```suggestion
const getLabelTitle = label => {
```

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+
+function getTitle (label) {
+    const prefix = getPrefix(label);
+    const type = getType(label);
```
With above suggestion:

```suggestion
    const { group, type } = parseLabel(label);
```

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+    const type = getType(label);
+
+    switch(prefix) {
+        case prefixes.FEATURE:
```
IMO this is why using capital constants is good idea; stupid people like me will read these two variables as the same (I did)!

nmadruga replied:
😸

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+        case prefixes.FEATURE:
+            switch (type) {
+                case featureTypes.EXISTING:
+                    return `\n\n## Improvements`;
```
Don't think we should return formatting in the string. Leave formatting to the caller IMO. E.g.:

```suggestion
                    return `Improvements`;
```

nmadruga replied:
I'll be doing this and the next one on a single commit.

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+        case prefixes.FEATURE:
+            switch (type) {
+                case featureTypes.EXISTING:
+                    return `\n\n## Improvements`;
```
Think these return strings all should probably be constants?

nmadruga replied:
Done

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+function checkSomeLabelEquals(labels, prefix, value) {
```
```suggestion
const checkSomeLabelEquals = (labels, prefix, value) => {
```

wlthomson replied:
I think this function is a bit confusing? AFAICS, this is checking if any label in `labels` matches a given prefix and value pair? I think would be clearer if use consistent naming, e.g. `labelGroup` and `labelType`?

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+    return labels.some(({name}) => name === `${prefix}: ${value}`);
+}
+
+function checkSomeLabelHasPrefix(labels, prefix) {
```
```suggestion
const checkSomeLabelHasPrefix = (labels, prefix) => {
```

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+    return labels.some(({name}) => name === `${prefix}: ${value}`);
+}
+
+function checkSomeLabelHasPrefix(labels, prefix) {
```
As above, bit confusing function name. Looks like this is checking is any label in `labels` belongs to a given `prefix` (i.e. group)? Maybe something like:

```suggestion
const checkSomeLabelHasGroup = (labels, group) => {
```

nmadruga replied:
Will make this change as part of the next suggestion on single commit

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+
+function checkSomeLabelHasPrefix(labels, prefix) {
+    labels.some(({name}) => getPrefix(name) === prefix);
+}
```
I think these functions could be made a bit more generic and easier to understand if they just took a single label? E.g.

```
const isLabelGroup = (label, group) => {
  const { group: labelGroup } = parseLabel(label);
  return labelGroup === group;
}
```

Caller can then use this with `some`, e.g. `labels.some(label => isLabelGroup(label, group));`

nmadruga replied:
Done. Thanks, that looks better.

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+}
+
+function checkSomeLabelHasPrefix(labels, prefix) {
+    labels.some(({name}) => getPrefix(name) === prefix);
```
Instead of `getPrefix`:

```suggestion
    labels.some(label => { 
      const { group: labelGroup } = parseLabel(label);
      return labelGroup === group;
    });
```

wlthomson commented on constants/labels.js:
```diff
@@ -0,0 +1,64 @@
+    labels.some(({name}) => getPrefix(name) === prefix);
+}
+
+module.exports = {
```
As above, might want to use `export` here.

wlthomson commented on githubAPI/octokit.js:
```diff
@@ -0,0 +1,58 @@
+if(!octokit.auth) {
+  console.log('Missing authentication token. Please add correct environment variable to file .env!');
+  return undefined;
+}
```
I think this should probably be wrapped in a function (e.g. a singleton constructor)?

nmadruga replied:
You just asks for difficult things! I have no idea how to make a singleton in JS> Ask me to make it in C++ and I'll be happy to do it. : D

wlthomson commented on githubAPI/octokit.js:
```diff
@@ -0,0 +1,58 @@
+
+if(!octokit.auth) {
+  console.log('Missing authentication token. Please add correct environment variable to file .env!');
+  return undefined;
```
I think good practice would be log a warning, but return the `octokit` object as is? IMO better to defer to the caller for validation; passes the responsibility down to the caller but also provides more flexibility to how the caller might want to use the object.

nmadruga replied:
Indeed. I'll make this change with in a single commit within the next suggested changes.

wlthomson commented on githubAPI/octokit.js:
```diff
@@ -0,0 +1,58 @@
+  const issuesAndPRs = await octokit.paginate(options);
+  const onlyIssues = issuesAndPRs.filter( issue => (!issue.pull_request))
+  return onlyIssues;
+}
```
For consistency, I think define the functions separately from `module.exports`;

```suggestion
const asyncTryToken = async () => {
  let checkToken = false;
  await octokit.repos.get({
    owner: process.env.OWNER,
    repo: process.env.REPO,
    mediaType: {
      previews: ["symmetra"]
    }
  })
  .catch( e => {
    console.log(e);
    return checkToken;
  });
}

const getIssues = async (params) => {
  const { milestone, filter, state, owner, repo } = params;

  await octokit.registerEndpoints({
    misc: {
      getRoot: {
        method: "GET",
        url: "/"
      }
    }
  });

  const options = octokit.issues.listForRepo.endpoint.merge({
    owner: owner ? owner : process.env.OWNER,
    repo: repo ? repo : process.env.REPO,
    milestone,
    state,
    labels: filter
  });

  const issuesAndPRs = await octokit.paginate(options);
  const onlyIssues = issuesAndPRs.filter( issue => (!issue.pull_request))
  return onlyIssues;
}

module.exports = { asyncTryToken, getIssues };
```

nmadruga replied:
Will be added as part of single commit

wlthomson commented on githubAPI/octokit.js:
```diff
@@ -0,0 +1,58 @@
+    console.log(e);
+    return checkToken;
+  });
+
```
Nit: whitespace.

nmadruga replied:
Will be added as part of single commit

wlthomson commented on githubAPI/octokit.js:
```diff
@@ -0,0 +1,58 @@
+  return undefined;
+}
+
+module.exports.asyncTryToken = async function() {
```
I think no need to specify `async` in the function identifier? Also I think good where possible to follow the convention is beginning boolean methods/variables with `is`. E.g, 

```suggestion
const isValidToken = async () => {
```

nmadruga replied:
Added as part of single commit

wlthomson commented on githubAPI/octokit.js:
```diff
@@ -0,0 +1,58 @@
+  return undefined;
+}
+
+module.exports.asyncTryToken = async function() {
```
As above wrt. removing globals, I think would be good to have this function called with the token, e.g.:

```suggestion
const isValidToken = async (token) => {
```

wlthomson commented on githubAPI/octokit.js:
```diff
@@ -0,0 +1,58 @@
+  });
+
+  const issuesAndPRs = await octokit.paginate(options);
+  const onlyIssues = issuesAndPRs.filter( issue => (!issue.pull_request))
```
```suggestion
  const issues = issuesPullRequests.filter( issue => (!issue.pull_request))
```

wlthomson commented on githubAPI/octokit.js:
```diff
@@ -0,0 +1,58 @@
+    labels: filter
+  });
+
+  const issuesAndPRs = await octokit.paginate(options);
```
IMO the `And` isn't needed here. Also, I think `pullRequest` instead of abbreviation `PR`:

```suggestion
  const issuesPullRequests = await octokit.paginate(options);
```

nmadruga replied:
Nice. I like your inputs a lot. Thanks @wlthomson very clean code practices.

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+const { getTitle } = require('../constants');
+
+let issuesNumbersIncluded = [];
```
I think this should be passed around as a local variable? Gobal variables are not a good idea IMO.

nmadruga replied:
Done

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+
+let issuesNumbersIncluded = [];
+
+function checkForDuplication(issue, duplicateIssues) {
```
```suggestion
const checkForDuplication = (issue, duplicateIssues) => {
```

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+let issuesNumbersIncluded = [];
+
+function checkForDuplication(issue, duplicateIssues) {
+  if(duplicateIssues === true) return `\n- ${issue.title} [${issue.number}]`;
```
```suggestion
  if (duplicateIssues) return `\n- ${issue.title} [${issue.number}]`;
```

nmadruga replied:
I will add this as part of a single commit

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+function checkForDuplication(issue, duplicateIssues) {
+  if(duplicateIssues === true) return `\n- ${issue.title} [${issue.number}]`;
+
+  const duplicate = issuesNumbersIncluded.some(numberIncluded => (numberIncluded === issue.number));
```
```suggestion
  const duplicate = issuesNumbersIncluded.some(numberIncluded => numberIncluded === issue.number);
```

nmadruga replied:
I will add this as part of a single commit

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+function checkForDuplication(issue, duplicateIssues) {
+  if(duplicateIssues === true) return `\n- ${issue.title} [${issue.number}]`;
+
+  const duplicate = issuesNumbersIncluded.some(numberIncluded => (numberIncluded === issue.number));
```
As prevous, IMO good convention is to prefix boolean variables with `is`:

```suggestion
  const isDuplicate= issuesNumbersIncluded.some(numberIncluded => numberIncluded === issue.number);
```

nmadruga replied:
I will add this as part of a single commit

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+
+let issuesNumbersIncluded = [];
+
+function checkForDuplication(issue, duplicateIssues) {
```
This function name is a bit unclear? Maybe something like `isDuplicateIssue`. IMO good to keep functions to answering a single question or performing a single action, suggest returning a boolean from this function and let the caller function decide what to do.

nmadruga replied:
Agree. I'll make this change as part of a single commit

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+  }
+}
+
+module.exports.generateChangesLog = function (groupIssues, {duplicateIssues, milestone}) {
```
As previous, good to keep consistency and define the function outside of `module.exports`:

```suggestion
const generateChangesLog = (groupIssues, {duplicateIssues, milestone}) => {
```

nmadruga replied:
will add this as part of a single commit

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+  }
+}
+
+module.exports.generateChangesLog = function (groupIssues, {duplicateIssues, milestone}) {
```
I think good to stay consistent on parameters. I prefer ES6-style object parameters, but I think combining them with comma-separated parameters is a bit confusing; best to chose one style or the other IMO.

nmadruga replied:
Choosing ES6-style!

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+module.exports.generateChangesLog = function (groupIssues, {duplicateIssues, milestone}) {
+  let changesLog = `# Changes Log for Milestone ${milestone}`;
```
IMO `let` is a bit of a code smell and should be avoided if possible. If go with above suggestion, would be easier to build up an array of objects here using e.g. `reduce`.

nmadruga replied:
Hmmm

wlthomson commented on helpers/changesLogGenerator.js:
```diff
@@ -0,0 +1,34 @@
+  }
+}
+
+module.exports.generateChangesLog = function (groupIssues, {duplicateIssues, milestone}) {
```
I think this function does a bit too much? At the moment the program data is quite tightly bound to the output, which I think will make it harder to make the log more flexible if we need to. Would be better IMO to build up a data structure and then parse that data structure into human-readable format in a separate function?

nmadruga replied:
@wlthomson Maybe this relates to the previous suggestion. Could you give more details on that one please? I'm a little bit clueless of how to do it without taking too long!

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+
+    const filterIssuesForAnyCustomer = function (issues) {
+      return issues.filter( issue => !checkSomeLabelHasPrefix(issue.labels, prefixes.CUSTOMER));
+    }
```
Do these functions need to be in function scope? Would be easier to read this function if they were in global scope?

nmadruga replied:
Done

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+const { prefixes, checkSomeLabelEquals, checkSomeLabelHasPrefix } = require('../constants');
+
+async function fetchIssuesInMilestoneByFilters (octokit, params) {
```
```suggestion
const fetchIssuesInMilestoneByFilters = async (octokit, params) {
```

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+    else if(customer){
+      return {
+        forCustomer: filterIssuesForCustomer(issues),
+        noCustomer: filterIssuesForAnyCustomer(issues)
```
These key's are a bit confusing. What do they mean?

nmadruga replied:
forCustomer = issues found with the label for the customer specified
noCustomer = issues found without a customer specified (which means a feature for any customer)

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+    else return { noCustomer: issues };
+}
+
+async function fetchIssuesByFilter(octokit, params, group){
```
```suggestion
const fetchIssuesByFilter = async (octokit, params, group) => {
```

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+
+async function fetchIssuesByFilter(octokit, params, group){
+  params.filter = group;
+  return await fetchIssuesInMilestoneByFilters(octokit, params);
```
This is mutating parameters :cry: . Best to clone them. Nice way to do this is with spread syntax:

```suggestion
  const issueParams = { ...params, filter: group };
  return await fetchIssuesInMilestoneByFilters(octokit, issueParams);
```

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+  return await fetchIssuesInMilestoneByFilters(octokit, params);
+}
+
+async function fetchIssuesUsingParams(octokit, params) {
```
```suggestion
const fetchIssuesUsingParams = async (octokit, params) => {
```

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+    customer: params.customer,
+    issues: await fetchIssuesInMilestoneByFilters(octokit, params)
+  });
+  return allIssues;
```
Does this need to return a singleton array? If so, can tidy this up a bit:

```suggestion
  const { customer } = params;
  const issues = await fetchIssuesInMilestoneByFilters(octokit, params);
  return [ { customer, issues } ]);
```

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+  return allIssues;
+}
+
+async function asyncForEach (array, octokit, params, callback) {
```
```suggestion
const asyncForEach = (array, octokit, params, callback) => {
```

nmadruga replied:
This is missing the async. I'll commit as part of a single commit with other changes in this module

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+  return allIssues;
+}
+
+async function asyncForEach (array, octokit, params, callback) {
```
These function and variable identifiers are a bit generic? What do `array`, `params` and `callback` represent?

nmadruga replied:
I'll rename the variables

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+      issues: await callback(octokit, params, array[index].trimLeft())
+    });
+  }
+  return groupedIssues;
```
IMO this can be done more concisely with `.map`:

```suggestion
  const { customer } = params;
  const issues = array.map(issue => issue.trimLeft());
  return issues.map(issue => ({
    key: issue,
    customer,
    issues: await callback(octokit, params, issue),
  });

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+  });
+
+  return issues;
+}
```
As previously:

```suggestion
const fetchIssuesInMilestone = async (octokit, params, filters) => {
  const issues = (!filters) ? await fetchIssuesUsingParams(octokit, params)
   : await asyncForEach(filters, octokit, params, fetchIssuesByFilter);

   issues.forEach(group => {
    const { forCustomer, noCustomer } = group.issues;
    (group.key) ? console.log('\nIssues type:', group.key) : console.log('\n');
    if(group.customer){
      console.log('No costumer:', noCustomer.length);
      console.log('For', group.customer, ':', forCustomer.length);
    } 
    else {
      console.log('Issues count:', noCustomer.length);
    }
  });

  return issues;
}

module.exports = { fetchIssuesInMilestone };
```

nmadruga replied:
All different now so, I'll add as part of a single commit to this module : ) Thanks!

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+
+module.exports.fetchIssuesInMilestone = async function (octokit, params, filters) {
+  const issues = (!filters) ? await fetchIssuesUsingParams(octokit, params)
+   : await asyncForEach(filters, octokit, params, fetchIssuesByFilter);
```
I think the negation here makes it a bit unnecessarily confusing?:

```suggestion
  const issues = filters ?
    await asyncForEach(filters, octokit, params, fetchIssuesByFilter) :
    await fetchIssuesUsingParams(octokit, params);
```

nmadruga replied:
Sure. I agree with all that. I don't know why I did like that in the first place 😭

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+   : await asyncForEach(filters, octokit, params, fetchIssuesByFilter);
+
+   issues.forEach(group => {
+    const { forCustomer, noCustomer } = group.issues;
```
Good to use deconstructing consistently:

```suggestion
    const { issues, key, customer } = group;
    const { forCustomer, noCustomer } = issues;
```

nmadruga replied:
Done

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+
+   issues.forEach(group => {
+    const { forCustomer, noCustomer } = group.issues;
+    (group.key) ? console.log('\nIssues type:', group.key) : console.log('\n');
```
Hmm, not sure why but this feels wrong to me!

```suggestion
    console.log(key ? `\nIssues type: ${key}` : '\n';
```

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+   issues.forEach(group => {
+    const { forCustomer, noCustomer } = group.issues;
+    (group.key) ? console.log('\nIssues type:', group.key) : console.log('\n');
+    if(group.customer){
```
```suggestion
    if (customer) {
```

nmadruga replied:
Done

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+    (group.key) ? console.log('\nIssues type:', group.key) : console.log('\n');
+    if(group.customer){
+      console.log('No costumer:', noCustomer.length);
+      console.log('For', group.customer, ':', forCustomer.length);
```
```suggestion
      console.log(`For ${customer}: ${forCustomer.length}`);
```

wlthomson commented on helpers/issuesFilter.js:
```diff
@@ -0,0 +1,71 @@
+      console.log('For', group.customer, ':', forCustomer.length);
+    } 
+    else {
+      console.log('Issues count:', noCustomer.length);
```
```suggestion
      console.log(`Issues count: ${noCustomer.length}`);
```

mark-prins commented:
Hi Nicole, I tried to run locally and `Error: Cannot find module './logs'`
I can't find that package.. what am I missing?

also - could you add these to .gitignore?
```
.DS_Store
yarn.lock
```

nmadruga commented:
@mark-prins Sorry! I didn't see your message until now! I forgot to commit the file...

Now it's commited. Try again please : )

mark-prins commented on constants/labels.js:
```diff
@@ -0,0 +1,65 @@
+function getTitle (label) {
+    const { group, type } = parseLabel(label);
+
+    switch(prefix) {
```
should this be:

```suggestion
    switch(label) {
```

and then the `prefix` below be changed as well?

mark-prins commented:
Hi @nmadruga - one more comment.. would you mind changing the value of `LIST` in `/constants/logs.js' to be `'* '`?

Then I can do a straight copy-n-paste into textpattern when generating the version history page 😄 

The asterisk works the same as hyphen in the github markdown - creates a bullet list too!

nmadruga commented:
> Hi @nmadruga - one more comment.. would you mind changing the value of `LIST` in `/constants/logs.js' to be `'* '`?
> 
> Then I can do a straight copy-n-paste into textpattern when generating the version history page 😄
> 
> The asterisk works the same as hyphen in the github markdown - creates a bullet list too!

Yes, that was exactly the goal, make a fila that can be just copy and paste to the version history and also can be posted to the Release notes on Github when the omsupply-bot starts using it.

I'll add your suggestion right now.

nmadruga commented:
@wlthomson When you have some time, please have a lock on the logic, and the not addressed suggestions. I had a reason to not do some, but others just seem a lot of effort now. Let's see what do you think, and I have accepted most of the suggestion. Hope it's looking tidier.

mark-prins commented:
it worked beautifully!! thanks Nicole 😀 Such a joy to use, and it makes the release notes a simple process.

Is it time to merge this PR and make new issue(s) for the remainder of the items, if there are any?
It might make the changes clearer / easier to follow

wlthomson commented:
I agree with @mark-prins that best to get this merged now. When I have some free time I will do a second pass over it, but in general looks good!
